{
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "7e34K-S7fqUr"
      },
      "outputs": [],
      "source": [
        "import random\n",
        "import copy\n",
        "from random import randint\n",
        "import matplotlib.pyplot as plt\n",
        "import openpyxl\n",
        "from openpyxl import Workbook\n",
        "from openpyxl.chart import LineChart, Reference\n",
        "import numpy as np\n",
        "\n",
        "#export to json"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "CY4hKrkcf33U"
      },
      "outputs": [],
      "source": [
        "genomesize = 102\n",
        "RNAcharacters = 'UGCA'\n",
        "#mutationProb = 2\n",
        "randomIndividuals = []\n",
        "\n",
        "class individual:\n",
        "  def __init__(self, generation):       #constructor - constructs a new individual object\n",
        "    self.generation = generation\n",
        "    self.fitness = 0\n",
        "    self.genome = []  #going to be an array of values eventually so define an empty array to begin\n",
        "    self.decoded_genome = None\n",
        "    for i in range(0,genomesize):\n",
        "      self.genome.append(random.choice(RNAcharacters))  #pulls elements from DNAcharacter string\n",
        "    self.calcFitness()\n",
        "    #self.color = (randint(0,255), randint(0,255), randint(0,255))\n",
        "  \n",
        "  def print(self):\n",
        "    for c in self.genome:\n",
        "      print(c, end = \"\")\n",
        "    print(\" \")\n",
        "    #print(self.color)\n",
        "    print(\"fitness:\" + str(self.fitness))\n",
        "\n",
        "    #plt.plot(self.fitness)\n",
        "    #print(type(self.genome))\n",
        "    \n",
        "  def fillArr(self, a):\n",
        "    a.append(self.fitness)\n",
        "    #print(a)\n",
        "  \n",
        "  def calcFitness(self):    \n",
        "    size = len(self.genome) - 2\n",
        "    self.fitness = 0\n",
        "    for c in range(0, size, 4):\n",
        "      if (self.genome[c] == 'C' and self.genome[c+1] == 'A' and self.genome[c+2] == 'U') or (self.genome[c] == 'C' and self.genome[c+1] == 'A' and self.genome[c+2] == 'C') :\n",
        "        #print(\"His\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'U' and self.genome[c+1] == 'U' and self.genome[c+2] == 'A') or (self.genome[c] == 'U' and self.genome[c+1] == 'U' and self.genome[c+2] == 'G') or (self.genome[c] == 'C' and self.genome[c+1] == 'U' and self.genome[c+2] == 'U') or (self.genome[c] == 'C' and self.genome[c+1] == 'U' and self.genome[c+2] == 'C')  or (self.genome[c] == 'C' and self.genome[c+1] == 'U' and self.genome[c+2] == 'A') or (self.genome[c] == 'C' and self.genome[c+1] == 'U' and self.genome[c+2] == 'G') :\n",
        "        #print(\"Leu\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'A' and self.genome[c+1] == 'A' and self.genome[c+2] == 'A') or (self.genome[c] == 'A' and self.genome[c+1] == 'A' and self.genome[c+2] == 'G') :\n",
        "        #print(\"Lys\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'A' and self.genome[c+1] == 'U' and self.genome[c+2] == 'G'):\n",
        "        #print(\"Met\")\n",
        "        self.fitness += 10\n",
        "      elif (self.genome[c] == 'A' and self.genome[c+1] == 'U' and self.genome[c+2] == 'U') or (self.genome[c] == 'A' and self.genome[c+1] == 'U' and self.genome[c+2] == 'C') or (self.genome[c] == 'A' and self.genome[c+1] == 'U' and self.genome[c+2] == 'A') :\n",
        "        #print(\"Iso\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'U' and self.genome[c+1] == 'U' and self.genome[c+2] == 'U') or (self.genome[c] == 'U' and self.genome[c+1] == 'U' and self.genome[c+2] == 'C') :\n",
        "        #print(\"Phe\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'A' and self.genome[c+1] == 'C' and self.genome[c+2] == 'U') or (self.genome[c] == 'A' and self.genome[c+1] == 'C' and self.genome[c+2] == 'C') or (self.genome[c] == 'A' and self.genome[c+1] == 'C' and self.genome[c+2] == 'A') or (self.genome[c] == 'A' and self.genome[c+1] == 'A' and self.genome[c+2] == 'G'):\n",
        "        #print(\"Thr\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'U' and self.genome[c+1] == 'G' and self.genome[c+2] == 'G'):\n",
        "        #print(\"Trp\")\n",
        "        self.fitness += 8\n",
        "      elif (self.genome[c] == 'G' and self.genome[c+1] == 'U' and self.genome[c+2] == 'U') or (self.genome[c] == 'G' and self.genome[c+1] == 'U' and self.genome[c+2] == 'C') or (self.genome[c] == 'G' and self.genome[c+1] == 'U' and self.genome[c+2] == 'A') or (self.genome[c] == 'G' and self.genome[c+1] == 'U' and self.genome[c+2] == 'G'):\n",
        "        #print(\"Val\")\n",
        "        self.fitness += 8\n",
        "      else:\n",
        "        self.fitness += 0\n",
        "\n",
        "    #for c in range(0, size):\n",
        "      #if self.genome[c] == 'A':\n",
        "        #self.fitness += 1\n",
        "\n",
        "  def mutation(self, mutationProb):\n",
        "    for i in range(0, genomesize):\n",
        "      if(random.uniform(0,100) < mutationProb):\n",
        "        self.genome[i] = random.choice(RNAcharacters)\n",
        "    self.calcFitness()\n",
        "\n",
        "  def copy(self, source):       #CALL = i2.copy(i1)\n",
        "    self.fitness = source.fitness\n",
        "    for i in range(0, genomesize):\n",
        "      self.genome[i] = source.genome[i]\n",
        "    self.generation = source.generation\n",
        "    \n",
        "  def __str__(self):\n",
        "    output = \"\"\n",
        "    for c in self.genome:\n",
        "        output = output + c\n",
        "    \n",
        "    if self.decoded_genome != None:\n",
        "      output += \" \"\n",
        "      for c in self.decoded_genome:\n",
        "        output = output + c\n",
        "    output += \" \"\n",
        "    output += \"fitness: \"\n",
        "    output += str(self.fitness)\n",
        "    output += \" \"\n",
        "    output += \"gen: \"\n",
        "    output += str(self.generation)\n",
        "    return output\n",
        "\n",
        "popsize = 22\n",
        "class population:\n",
        "  def __init__(self, parent = None):       #constructor - constructs a new population object\n",
        "    #print(\"Creating a new population\") \n",
        "    self.current_gen = 0\n",
        "    self.avgFitness = 0\n",
        "    self.the_pop = []  \n",
        "    if parent is None:\n",
        "      for i in range(0, popsize):\n",
        "        self.the_pop.append(individual(self.current_gen))  #pulls elements from DNAcharacter string\n",
        "    else:\n",
        "      for i in range(0, popsize):\n",
        "        i1 = individual(self.current_gen)\n",
        "        i1.copy(parent)\n",
        "        self.the_pop.append(i1)\n",
        "\n",
        "  def calcStats(self):\n",
        "    self.avg_fitness = 0\n",
        "    for i in self.the_pop:\n",
        "      self.avg_fitness += i.fitness\n",
        "    self.avg_fitness /= popsize\n",
        "\n",
        "  def generational(self, n, t):\n",
        "    self.current_gen += 1\n",
        "    if self.current_gen % 3 == 1:\n",
        "      randomIndividuals.append(\n",
        "        \n",
        "        copy.deepcopy((self.the_pop[randint(0,popsize-1)]))\n",
        "        )\n",
        "    tempPop = population()\n",
        "    for i in range(0, popsize, 2):    #needs an even pop size\n",
        "      parent = self.tournament(t) #select, returns an index\n",
        "      parent2 = self.tournament(t)\n",
        "      tempPop.the_pop[i].copy(self.the_pop[parent])    \n",
        "      tempPop.the_pop[i+1].copy(self.the_pop[parent2])\n",
        "      tempPop.crossover(i, i+1)\n",
        "      tempPop.the_pop[i+1].mutation(n)\n",
        "      tempPop.the_pop[i].mutation(n)\n",
        "      tempPop.the_pop[i].generation = self.current_gen\n",
        "      tempPop.the_pop[i+1].generation = self.current_gen\n",
        "    #select individuals using tournament()\n",
        "    #mutate them?\n",
        "    #put into new/temp population \n",
        "\n",
        "    #when new/temp population is full, copy new/temp pop back into the_pop\n",
        "    for i in range(0, popsize):\n",
        "      self.the_pop[i].copy(tempPop.the_pop[i])\n",
        "    self.calcStats()\n",
        "\n",
        "  # def onepoint_crossover(self,p1,p2):   #pick random point and cross anything beyond that point\n",
        "  #   crossover_point = random.randint(0, genomesize)\n",
        "  #   for j in range(crossover_point, genomesize):\n",
        "  #     temp = self.the_pop[p1].genome[j] #remeber parent 1's character\n",
        "  #     self.the_pop[p1].genome[j] = self.the_pop[p2].genome[j]\n",
        "  #     self.the_pop[p2].genome[j] = temp\n",
        "\n",
        "\n",
        "  def crossover(self,p1,p2):    #uniform crossover- 10% possibility of swaping any two characters\n",
        "    for j in range(0, genomesize):\n",
        "      if random.randint(0, 100) < 10:       #uniform crossover\n",
        "          #print(j)\n",
        "          temp = self.the_pop[p1].genome[j] #remeber parent 1's character\n",
        "          self.the_pop[p1].genome[j] = self.the_pop[p2].genome[j]\n",
        "          self.the_pop[p2].genome[j] = temp\n",
        "          #self.the_pop[p1].genome[j],self.the_pop[p2].genome[j],self.the_pop[p1].genome\n",
        "          #a,b = b,a\n",
        "  def tournament(self, t):\n",
        "    tourn_size = t\n",
        "    best_so_far = random.randint(0, popsize-1)\n",
        "    best_fitness = self.the_pop[best_so_far].fitness\n",
        "    #print(best_so_far)\n",
        "    for i in range(0, tourn_size - 1):\n",
        "      current = random.randint(0, popsize-1)\n",
        "      current_fit = self.the_pop[current].fitness\n",
        "      #print(current)\n",
        "      if(current_fit > best_fitness):\n",
        "        best_so_far = current\n",
        "        best_fitness = current_fit\n",
        "    return best_so_far"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "o_D8ertV7cwN"
      },
      "outputs": [],
      "source": [
        "def decode(s):  # decodes each codon into a single char amino acid\n",
        "    string = \"\"\n",
        "    for char in s:\n",
        "        string += char\n",
        "    # print(s)\n",
        "    match string:\n",
        "        case 'AAA' | 'AAG': #lys\n",
        "            return 'k'\n",
        "        case 'AAU' | 'AAC': # Asn\n",
        "            return 'n'\n",
        "        case 'ACU' | 'ACC' | 'ACA' | 'ACG': #Thr\n",
        "            return 't'\n",
        "        case 'AGU' | 'AGC': # Ser\n",
        "            return 's'\n",
        "        case 'AUU'|'AUC'|'AUA': # iie\n",
        "            return 'i'\n",
        "        case 'AUG': # met and start\n",
        "            return 'm'\n",
        "        case 'AGA' | 'AGG': # arg\n",
        "            return 'r'\n",
        "        case 'CGA' | 'CGC' | 'CGG' | 'CGU': # arg\n",
        "            return 'r'\n",
        "        case 'CAU' | 'CAC': # His\n",
        "            return 'h'\n",
        "        case 'CAA' | 'CAG': # Gln\n",
        "            return 'q'\n",
        "        case 'CCA' | 'CCC' | 'CCG' | 'CCU': # Pro\n",
        "            return 'p'\n",
        "        case 'CUA' | 'CUC' | 'CUG' | 'CUU': # Leu\n",
        "            return 'l'\n",
        "        case 'GUA' | 'GUC' | 'GUG' | 'GUU': # Val\n",
        "            return 'v'\n",
        "        case 'GCA' | 'GCC' | 'GCG' | 'GCU': # Ala\n",
        "            return 'a'\n",
        "        case 'GAU' | 'GAC': # Asp \n",
        "            return 'd'\n",
        "        case 'GAG' | 'GAA': # Gku\n",
        "            return 'e'\n",
        "        case 'GGA' | 'GGC' | 'GGG' | 'GGU': # Gly\n",
        "            return 'g'\n",
        "        case 'UUU' | 'UUC': # phe\n",
        "            return 'f'\n",
        "        case 'UUA' | 'UUG': # Leu\n",
        "            return 'l'\n",
        "        case 'UCA' | 'UCC' | 'UCG' | 'UCU': # Ser\n",
        "            return 's'\n",
        "        case 'UAU' | 'UAC': # Tyr \n",
        "            return 'y'\n",
        "        case 'UAA'|'UAG'|'UGA': # Stop \n",
        "            return '!'\n",
        "        case 'UGU' | 'UGC': # cys\n",
        "            return 'c'\n",
        "        case 'UGG': # Trp,  \n",
        "            return 'w'\n",
        "        case _:\n",
        "            # print(\"Unidentified codon\")\n",
        "            return '?'\n",
        "\n",
        "# print(decode(['A','A','A']))\n",
        "\n",
        "def decode_genome(individual):\n",
        "    i = 0\n",
        "    length = len(individual.genome)\n",
        "    decoded_genome = []\n",
        "    while i < length:\n",
        "        decoded_protein = decode(individual.genome[i:i+3])\n",
        "        decoded_genome.append(decoded_protein)\n",
        "        i += 3\n",
        "    individual.decoded_genome = decoded_genome"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "wjsJzaw-ha3p",
        "outputId": "5a6eebcc-7bca-4e27-dce5-e9f86ee63524"
      },
      "outputs": [],
      "source": [
        "p = population(individual(0))\n",
        "n = 0.2\n",
        "t = 5\n",
        "x  = []\n",
        "numGens = 2000\n",
        "\n",
        "for i in range(1, numGens+1):\n",
        "  x.append(i)\n",
        "\n",
        "#for i in range(0, popsize):\n",
        "  #p.the_pop[i].print()\n",
        "  #print(\"hi\")\n",
        "\n",
        "#for l in range(3):\n",
        "a = []\n",
        "print(n)\n",
        "for i in range(0, numGens):\n",
        "  p.generational(n, t)\n",
        "  #p.the_pop[0].print()\n",
        "  p.the_pop[0].fillArr(a)\n",
        "\n",
        "\n",
        "plt.plot(x, a, label=\"0.2 Mutation Rate\")\n",
        "\n",
        "\n",
        "plt.plot\n",
        "\n",
        "plt.xlabel(\"Generation\")\n",
        "plt.ylabel(\"Fitness\")\n",
        "plt.title(\"Fitness vs Mutation Rate\")\n",
        "plt.legend()\n",
        "plt.show()\n"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "o_OQA8ot5kj9"
      },
      "outputs": [],
      "source": [
        "# print(randomIndividuals)\n",
        "for ind in randomIndividuals:\n",
        "    decode_genome(ind)\n",
        "    print(ind)"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "cF2JwuzB5pmX"
      },
      "outputs": [],
      "source": [
        "import json\n",
        "outputData = {}\n",
        "\n",
        "for ind in randomIndividuals:\n",
        "    outputData.update({str(ind.generation): [ind.genome, ind.fitness]})\n",
        "\n",
        "json_object = json.dumps(outputData, indent=4)\n",
        "\n",
        "with open(\"testData.json\", \"w\") as outfile:\n",
        "    outfile.write(json_object)"
      ]
    }
  ],
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "codemirror_mode": {
        "name": "ipython",
        "version": 3
      },
      "file_extension": ".py",
      "mimetype": "text/x-python",
      "name": "python",
      "nbconvert_exporter": "python",
      "pygments_lexer": "ipython3",
      "version": "3.10.4"
    }
  },
  "nbformat": 4,
  "nbformat_minor": 0
}
